snuba/query/processors/prewhere.py

```diff
@@ -25,56 +34,189 @@ class PrewhereProcessor(QueryProcessor):
+            :max_prewhere_conditions
+        ]
+
+        if prewhere_conditions:
+            self._update_conditions(query, prewhere_conditions)
+
+    @abstractmethod
+    def _get_prewhere_candidates(
+        self, query: Query, prewhere_keys: Sequence[str]
+    ) -> Sequence[Tuple[Iterable[TColumn], TCondition]]:
+        """
+        Extract the conditions from the query that could be moved to the
+        pre-where clause.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def _get_column_name(self, column: TColumn) -> str:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _update_conditions(
+        self, query: Query, prewhere_conditions: Sequence[TCondition]
+    ) -> None:
+        """
+        Updates the query with the new pre-where conditions by removing them
+        from the main condition clause and adding them to the pre-where clause.
+        """
+        raise NotImplementedError
+
+
+class LegacyPrewhereProcessor(PrewhereProcessorDelegate[Condition, str]):
+    def _get_prewhere_candidates(
+        self, query: Query, prewhere_keys: Sequence[str]
+    ) -> Sequence[Tuple[Iterable[str], Condition]]:
         # Add any condition to PREWHERE if:
         # - It is a single top-level condition (not OR-nested), and
         # - Any of its referenced columns are in prewhere_keys
         conditions = query.get_conditions()
         if not conditions:
-            return
-        prewhere_candidates = [
+            return []
+        return [
             (util.columns_in_expr(cond[0]), cond)
             for cond in conditions
             if util.is_condition(cond)
             and cond[1] in ALLOWED_OPERATORS
             and any(col in prewhere_keys for col in util.columns_in_expr(cond[0]))
         ]
-        # Use the condition that has the highest priority (based on the
-        # position of its columns in the prewhere keys list)
-        prewhere_candidates = sorted(
+
+    def _get_column_name(self, column: str) -> str:
+        return column
+
+    def _update_conditions(
+        self, query: Query, prewhere_conditions: Sequence[Condition]
+    ) -> None:
+        conditions = query.get_conditions()
+        if not conditions:
+            # This should never happen at this point, but for mypy this can be None.
+            return
+
+        query.set_conditions(
+            [cond for cond in conditions if cond not in prewhere_conditions]
+        )
+        query.set_prewhere(prewhere_conditions)
+
+
+class ASTPrewhereProcessor(PrewhereProcessorDelegate[Expression, Column]):
+    """
+    This class MUST not depend on anything from the legacy implementation because,
+    as long as we have both ast and legacy representation coexisting, this class
+    cannot be sure whether the legacy representation has already been modified
+    to process the pre-where conditions.
+    """
+
+    allowed_ast_operators = [OPERATOR_TO_FUNCTION[o] for o in ALLOWED_OPERATORS]
```

Comment:
I vaguely recall some conversation about further limiting the prewhere operators (e,g. not `!=`), did I remember correctly? What was the decision there?

Reply:
The issue with this approach as a whole is that we are making the decision whether to add the expression to the prewhere or not based on the operator alone, while this decision should be made on the selectiveness of the condition.
In order to add them to the prewhere condition if they are highly selective conditions only we would need stats around the table we do not have, so there is really no clear win by just depending on the operator.
The decision of keeping only NOT IN out of the list is a bit arbitrary but driven by the fact that we have a clear issue with that while we do not have such clear issue with `!=`

Reply:
Ok. Just curious, no reason to change as part of this anyway.